lib/backend/watchersyncer/watchercache.go

```diff
@@ -0,0 +1,285 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watchersyncer
+
+import (
+	"context"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	cerrors "github.com/projectcalico/libcalico-go/lib/errors"
+)
+
+type watcherCache struct {
+	logger       *logrus.Entry
+	client       api.Client
+	watch        api.WatchInterface
+	resources    map[string]resourceKey
+	oldResources map[string]resourceKey
+	results      chan<- interface{}
+	hasSynced    bool
+	resourceType ResourceType
+}
+
+type resourceKey struct {
+	revision string
+	key      model.Key
+}
+
+// Create a new watcherCache.
+// Note that the results channel is untyped.  The watcherSyncer expects one of the following
+// types:
+// -  An error
+// -  An api.Update
+// -  A api.SyncStatus (only for the very first InSync notification)
+func newWatcherCache(client api.Client, resourceType ResourceType, results chan<- interface{}) *watcherCache {
+	wc := &watcherCache{
+		logger:       logrus.WithField("ListRoot", model.ListOptionsToDefaultPathRoot(resourceType.ListInterface)),
+		client:       client,
+		resourceType: resourceType,
+		results:      results,
+		resources:    make(map[string]resourceKey, 0),
+	}
+	return wc
+}
+
+// run creates the watcher and loops indefinitely reading from the watcher.
+func (wc *watcherCache) run() {
+	wc.createWatcher()
+	for {
+		select {
+		case event := <-wc.watch.ResultChan():
```

Comment:
I think you should be able to simplify:
```go
for{
  select{
      case event := <-wc.watch.ResultChan():
          switch event.Type {
               api.WatchSynced:
 ...
``` 
to 
```go
for{
  switch <-wc.watch.ResultChan().Type {
       case api.WatchSynced:
...
```
because `select` `case` is usually if you're waiting on multiple channels, in this case we just have one channel we're waiting on. I tried this on a simple program, and it seemed to work.

Reply:
I don't think that's quite right, but you are correct that I could drop the select.

I think it would be more like the following, because I need access to event, not just event.Type:

```
event := <=wc.watch.ResultChan()
switch event.Type {
...
```